Improve memory safety in WebEditorClient.cpp

zakariaridouh opened the pull request:
#### 82d8972bf1c96622bd8bae7325dc889bf7be6898
<pre>
Improve memory safety in WebEditorClient.cpp
<a href="https://rdar.apple.com/159779944">rdar://159779944</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=298336">https://bugs.webkit.org/show_bug.cgi?id=298336</a>

Reviewed by Chris Dumez.

Fix failing safercpp issues in WebProcess/WebCoreSupport/WebEditorClient.cpp.

* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
(WebKit::WebEditorClient::textFieldDidEndEditing):
(WebKit::WebEditorClient::textDidChangeInTextField):
(WebKit::WebEditorClient::textDidChangeInTextArea):
(WebKit::WebEditorClient::doTextFieldCommandFromEvent):
(WebKit::WebEditorClient::textWillBeDeletedInTextField):

Canonical link: <a href="https://commits.webkit.org/300526@main">https://commits.webkit.org/300526@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/567e78bd11df868bbb9114dea08a5cac65cb3d7c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122767 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42479 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33169 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129392 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74871 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c1c4026a-d5a9-461b-8857-fc671e831881) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51072 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93305 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61939 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8e9d6c25-39ee-4b11-810e-a7a6f43f9c08) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125719 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34443 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109901 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73944 "Found 1 new API test failure: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlotTimeout (failure)") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7d13e483-490d-4c06-a5d5-f52ac5c5ae3a) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33428 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72882 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104143 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28268 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132119 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49712 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37843 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101819 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50089 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106115 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101691 "Found 2 new API test failures: TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25862 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47066 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25246 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46482 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49569 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55322 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49036 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52388 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50719 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4dd92f14b64a8fac45e4129144d75539568d49c7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50269

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
```diff
@@ -118,32 +118,32 @@ int WebEditorClient::spellCheckerDocumentTag()
 bool WebEditorClient::shouldBeginEditing(const SimpleRange& range)
 {
-    return m_page->injectedBundleEditorClient().shouldBeginEditing(m_page.get(), range);
+    return m_page->injectedBundleEditorClient().shouldBeginEditing(protectedPage->get(), range);
```
`protectedPage()`  ?

zakariaridouh replied:
Yup, accidentally pushed as non draft. Will switch back to ready for review soon.

zakariaridouh commented on Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
```diff
@@ -206,24 +206,24 @@ void WebEditorClient::didBeginEditing()
     // FIXME: What good is a notification name, if it's always the same?
     static NeverDestroyed<String> WebViewDidBeginEditingNotification(MAKE_STATIC_STRING_IMPL("WebViewDidBeginEditingNotification"));
-    m_page->injectedBundleEditorClient().didBeginEditing(m_page.get(), WebViewDidBeginEditingNotification.get().impl());
+    m_page->injectedBundleEditorClient().didBeginEditing(protectedPage().get(), WebViewDidBeginEditingNotification.get().impl());
```
Good idea.

zakariaridouh replied:
WebPage doesn't extend `CanMakeCheckedPtr`, have made it extend it. Is this the right way to go about converting it to a checkedRef?

cdumez replied:
I don't see WebPage extending `CanMakeCheckedPtr` in this latest patch iteration, EWS is also all red. But yes, I think having WebPage extend `CanMakeCheckedPtr` would be fine.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
```diff
@@ -435,7 +435,7 @@ void WebEditorClient::textFieldDidEndEditing(Element& element)
         return;
 
-    auto webFrame = WebFrame::fromCoreFrame(*element.document().frame());
+    auto webFrame = WebFrame::fromCoreFrame(Ref { *element.document().frame() }.get());
```
Should use `protectedFrame()` here and elsewhere in this PR.

zakariaridouh replied:
fixed !

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/567e78bd11df868bbb9114dea08a5cac65cb3d7c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50269

zakariaridouh committed:
Improve memory safety in WebEditorClient.cpp
rdar://159779944
https://bugs.webkit.org/show_bug.cgi?id=298336

Reviewed by Chris Dumez.

Fix failing safercpp issues in WebProcess/WebCoreSupport/WebEditorClient.cpp.

* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
(WebKit::WebEditorClient::textFieldDidEndEditing):
(WebKit::WebEditorClient::textDidChangeInTextField):
(WebKit::WebEditorClient::textDidChangeInTextArea):
(WebKit::WebEditorClient::doTextFieldCommandFromEvent):
(WebKit::WebEditorClient::textWillBeDeletedInTextField):

Canonical link: https://commits.webkit.org/300526@main

webkit-commit-queue commented:
Committed 300526@main (82d8972bf1c9): <https://commits.webkit.org/300526@main>

Reviewed commits have been landed. Closing PR #50269 and removing active labels.
